[SE-3381] Allows adding new tinymce plugins through platform configuration

nizarmah opened the pull request:
This makes it possible to install new tinymce plugins through the environment settings.

This PR is responsible for adding the tinymce plugin settings to the tinymce javascript configuration.

**JIRA tickets**: SE-3381, SE-3247, [OSPR-5263](https://openedx.atlassian.net/browse/OSPR-5263), [OSPR-5264](https://openedx.atlassian.net/browse/OSPR-5264)

**Dependencies**:

- https://github.com/edx/configuration/pull/6179

**Sandbox URL**: 

- LMS: https://pr25695.sandbox.opencraft.hosting/
- Studio: https://studio.pr25695.sandbox.opencraft.hosting/

**Testing instructions**:

1. Login to Studio using `staff@example.com`/`edx`
2. Create a new Unit with an HTML Text
3. Make sure that the *ADSK Link* shows in the toolbar
4. Create a link and make sure you can preview it, and see it on the LMS

**Author Notes and Concerns**:

1. This is a limited implementation, it does not support custom stylesheets in the TinyMCE plugins.
2. The reason behind that is that no plugin currently requires any stylesheets to be included into Studio and LMS.

**Reviewers**
- [x] @pkulkark 
- [x] @bradenmacdonald 
- [ ] edX reviewer[s] TBD

**Settings**
```yaml
EDXAPP_CMS_ENV_EXTRA:
  JS_ENV_EXTRA_CONFIG:
    TINYMCE_ADDITIONAL_PLUGINS:
    - name: adsklink
      toolbar: true
      extra_settings:
        linktypes:
        - download
        - offer
        filetypes:
        - PDF
        - zip
        - Video
        - Design
        orientations:
        - Vertical
        - Horizontal
        styles:
        - Primary
        - Normal
        - Secondary

TINYMCE_ADDITIONAL_PLUGINS_LIST:
- repo: https://gitlab.com/nizarmah/tinymce-adsk-plugin
  name: adsklink
  plugin_path: "/adsklink"

edx_ansible_source_repo: "https://github.com/open-craft/configuration.git"
configuration_version: "nizar/tinymce_plugins_role"
edx_platform_commit: "nizar/tinymce_plugins_combined_dependencies_master"
```

edx-webhook commented:
Thanks for the pull request, @nizarmah! I've created [OSPR-5264](https://openedx.atlassian.net/browse/OSPR-5264) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

natabene commented:
@nizarmah Thank you for your contribution. Please let me know once it is ready for a review.

nizarmah commented:
@natabene small note by the way, I know I have a bunch of pull requests open at the moment. I just wanted to let you know that even if I haven't made any changes on a pull request, I haven't abandoned any of them.

I'm keeping good track of them, and switching between them depending on the work's priority internally. :+1:

bradenmacdonald commented on common/lib/xmodule/xmodule/js/src/html/edit.js:
```diff
@@ -171,7 +172,37 @@
+        };
+
+        if (process.env.ADDITIONAL_NODE_ENV_VARS) {
+          var tinyMceAdditionalPlugins = JSON.parse(process.env.ADDITIONAL_NODE_ENV_VARS).TINYMCE_ADDITIONAL_PLUGINS;
```
This seems a bit strange. I would expect that if I set `ADDITIONAL_NODE_ENV_VARS = {"foo": "bar"}` in the platform configuration, then I would be able to read it as `process.env.foo`

In other words, I would expect that a key set in a variable called "additional node environment variables" would become a node environment variable.

Maybe just rename `ADDITIONAL_NODE_ENV_VARS` to `JS_ENV_EXTRA_CONFIG` or something?

nizarmah replied:
Yeah, you have a good point. Thanks for the suggestion which I went with, it made it easier to make the change. :+1:

nizarmah commented:
Apparently, all it took to fix the python build was 2 commits, rasied to the third power :slightly_smiling_face:

pkulkark reviewed (approved):
LGTM :+1: 

- [x] I tested this: verified the new tinymce plugin is installed correctly as described in the testing instructions
- [x] I read through the code

bradenmacdonald commented:
@nizarmah Could you please include documentation about this new extension point in [`extension_points.rst`](https://github.com/edx/edx-platform/blob/master/docs/guides/extension_points.rst) ? That will also notify people watching for changes that improve platform extensibility.

nizarmah commented on common/lib/xmodule/xmodule/js/src/html/edit.js:
```diff
@@ -171,7 +172,37 @@
+              if (tinyMceConfig.toolbar.trim()) {
+                tinyMceConfig.toolbar += ' | ';
+              }
+              tinyMceConfig.toolbar += plugin_name;
```
It has come to my attention that not all TinyMCE plugins will need to be present in the toolbar. I'll need to address that in a slight change.

nizarmah replied:
Alright, this has been resolved :+1:

nizarmah commented:
Alright, so what has been addressed is the following:
- Added information about the limitation of the solution in the pull request description
- Changed the method of passing the additional tinymce plugins configuration:
  - Now each plugin is passed as a dictionary, instead of a key
  - Each plugin has a `toolbar` key in order to indicate whether it should be added to the toolbar
  - Each plugin has `extra_settings` in order to hold its extra settings
- Also @bradenmacdonald  I have added a brief mention of the TinyMCE plugins in the [`extension_points.rst`](https://github.com/edx/edx-platform/blob/f801319fdd5a1f60c32a8d47064101ba4143ac2e/docs/guides/extension_points.rst). But I have not created an actual doc to include in edX's documentation. Would you prefer that I create one and link to it, instead of linking to this pull request?

nizarmah commented:
@bradenmacdonald sorry, I'll need you to hold slightly on reviewing this ticket. There's something that I should've checked for, and if this PR would break Multiple App Servers.

The reason is, that we're doing changes to `edit.js` which would be reflected in the bundled `common.js`, and if there's a slight sorting issue due to passing the `JS_ENV_EXTRA_CONFIG`, then they'll break.

So I'll be building multiple app servers to verify that the hash is identical 👍🏼

bradenmacdonald commented:
No problem @nizarmah, just let me know.

@natabene can I take this as a core committer?

natabene commented:
@bradenmacdonald Sure.

nizarmah commented:
@bradenmacdonald thanks a lot for being patient with me 😄 I ended up creating 5 different servers and made sure that the bundled `commons.js` is generated consistently among the different servers, without any single change (content hashes were identical on all machines 🥳)

So now that I've verified the change doesn't affect multiple app server support, this is ready for your review again 👍🏼

bradenmacdonald commented:
@nizarmah Unfortunately there seems to be an error on the sandbox - the visual editor is not displaying:

<img width="764" alt="Screen Shot 2020-12-18 at 1 14 20 PM" src="https://user-images.githubusercontent.com/945577/102662109-ffc1e700-4132-11eb-991b-1621a5524d3f.png">

bradenmacdonald commented on cms/envs/common.py:
```diff
@@ -1020,6 +1020,9 @@
 ##### EMBARGO #####
 EMBARGO_SITE_REDIRECT_URL = None
 
+##### custom vendor plugin variables #####
```
Perhaps say also that "JavaScript code can access this data using `JSON.parse(process.env.JS_ENV_EXTRA_CONFIG)`", and mention that one use case is enabling custom TinyMCE plugins.

nizarmah replied:
This has been addressed in [`e882cd4 - Adds more information about the JS_ENV_EXTRA_CONFIG env setting`](https://github.com/edx/edx-platform/pull/25695/commits/e882cd41a7c5a25f94bfe15d42fc478ed37469cd) 👍🏼

bradenmacdonald commented on docs/guides/extension_points.rst:
```diff
@@ -63,6 +63,9 @@ If you want to provide learners with new content experiences within courses, opt
      - Hold, Stable
      - An external grader is a service that receives learner responses to a problem, processes those responses, and returns feedback and a problem grade to the edX platform. You build and deploy an external grader separately from the edX platform. An external grader is particularly useful for software programming courses where learners are asked to submit complex code. See the `external grader do ...
+   * - **TinyMCE Plugins**
+     - Hold, Limited
```
I would say "Trial", not "hold".
"Hold" means it's legacy code that shouldn't be used anymore, and that shouldn't be true of something that we are just adding to the platform now :)

nizarmah replied:
Yeah, that was silly of me 😅 Thanks for clarifying that 😃 

This has been addressed in [`e6803fa - Updates the extension points and adds a guide on how to install custom tinymce plugins`](https://github.com/edx/edx-platform/pull/25695/commits/e6803fab65f0e35badf21addb214c46819b5fc91) 👍🏼

bradenmacdonald commented on docs/guides/extension_points.rst:
```diff
@@ -63,6 +63,9 @@ If you want to provide learners with new content experiences within courses, opt
    * - **External Graders**
      - Hold, Stable
      - An external grader is a service that receives learner responses to a problem, processes those responses, and returns feedback and a problem grade to the edX platform. You build and deploy an external grader separately from the edX platform. An external grader is particularly useful for software programming courses where learners are asked to submit complex code. See the `external grader do ...
+   * - **TinyMCE Plugins**
```
People might not know what TinyMCE is. This should say "TinyMCE (Visual Text/HTML Editor) Plugins"

nizarmah replied:
Yeah, that's a great remark. I added it in the guide I added too 🙂 

This has been addressed in [`e6803fa - Updates the extension points and adds a guide on how to install custom tinymce plugins`](https://github.com/edx/edx-platform/pull/25695/commits/e6803fab65f0e35badf21addb214c46819b5fc91) 👍🏼

bradenmacdonald commented on docs/guides/extension_points.rst:
```diff
@@ -72,6 +75,7 @@ For a more detailed comparison of content integration options, see `Options for
 .. _Options for Extending the edX Platform: https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/extending_platform/extending.html
 .. _custom JavaScript application: https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/extending_platform/javascript.html
 .. _external grader documentation: https://edx.readthedocs.io/projects/open-edx-ca/en/latest/exercises_tools/external_graders.html
+.. _the additional configuration settings for TinyMCE: https://github.com/edx/edx-platform/pull/25695
```
To answer your other question, yes it would be better if this can link to a .rst documentation file somewhere, rather than a pull request.

nizarmah replied:
Yeah, I created a new directory in the `docs/guides` in order to add any guide close to the `extension_points` you mentioned.

Let me know if you think there's a better place for it, please.

This has been addressed in [`e6803fa - Updates the extension points and adds a guide on how to install custom tinymce plugins`](https://github.com/edx/edx-platform/pull/25695/commits/e6803fab65f0e35badf21addb214c46819b5fc91) 👍🏼

bradenmacdonald commented on pavelib/assets.py:
```diff
@@ -778,10 +779,19 @@ def webpack(options):
         static_root_lms=static_root_lms,
-        static_root_cms=static_root_cms
+        static_root_cms=static_root_cms,
+        js_env_extra_config=js_env_extra_config
```
Nit: please add a trailing comma, to get nicer diffs in the future.

nizarmah replied:
Yeap, sorry about missing that 👍🏼 This should be addressed now 😃

bradenmacdonald commented on webpack.dev.config.js:
```diff
@@ -20,7 +20,8 @@ module.exports = _.values(Merge.smart(commonConfig, {
             new webpack.DefinePlugin({
-                'process.env.NODE_ENV': JSON.stringify('development')
+                'process.env.NODE_ENV': JSON.stringify('development'),
+                'process.env.JS_ENV_EXTRA_CONFIG': JSON.stringify(process.env.JS_ENV_EXTRA_CONFIG)
```
Nit: add trailing comma.

bradenmacdonald replied:
Also I think something weird is going on here?
`process.env.JS_ENV_EXTRA_CONFIG` is already a string, but you are calling `JSON.stringify` on it?

And above, you have `json.dumps(str({}))` which is a confusing way of writing `'"{}"'`, which looks like something that's doubly-encoded into JSON.

And above, `json.dumps(js_env_extra_config_setting.replace("'", '"'))` looks like of hacky.



Can this be simplified to something like:

```
js_env_extra_config_setting, = Env.get_django_settings(["JS_ENV_EXTRA_CONFIG"], "cms", settings=settings)
js_env_extra_config = json.dumps(js_env_extra_config_setting or {})
```
and here:

```
'process.env.JS_ENV_EXTRA_CONFIG': process.env.JS_ENV_EXTRA_CONFIG,  // Already a JSON string
```

nizarmah replied:
> Nit: add trailing comma.

Unfortunately, adding the trailing comma made the quality builds fail. Eslint returned an error for the trailing commas, so I went ahead and removed them as part of [`78a6cf1`](https://github.com/edx/edx-platform/pull/25695/commits/78a6cf1f72a9bab7842ef52d68536920af974815).

Sorry about that! 

---

Regarding to the second comment, about how hacky the solution is, I apologize for that.

It was careless of me to overlook those lines, just because they were "working". As you suggested, I changed the approach.

The changes you requested were done in [`441fcad - Simplifies the way js env extra config is being passed to webpack`](https://github.com/edx/edx-platform/pull/25695/commits/441fcad5fc0fed7d3442aa2654c0858648a95d35).

Those changes successfully passed the JSON String to Webpack, as you explained in your comment. However, there was another problem, which was causing the error in the sandbox that you [mentioned earlier](https://github.com/edx/edx-platform/pull/25695#issuecomment-748322795).

This problem was because of the JSON String that was being passed to Webpack, and JavaScript was attempting to parse, using `JSON.parse`. The string was shaped in the following way:
```json
"{ ... 'toolbar': True }"
```

Unfortunately, that's due to the `print_setting` that is being used to extract the settings from CMS and LMS.
Accordingly, a new comment had to be added: [`dabac7a - Re-formats js env extra config settings to valid json parseable string`](https://github.com/edx/edx-platform/pull/25695/commits/dabac7a12580dec92f3a2c1e37b8b551d8d0b9f9)

This simply changes the `True` and `False` values in the JSON string to `true` and `false` using regular expressions, which makes the JSON string valid. 

_However_, once that was done, Webpack started reading the JSON string as a parsed object 🙂 
So, I had to remove the parsing that was already done in JavaScript, reference to commit [`6fa07ec - Removes the need to parse the webpack env variable passed`](https://github.com/edx/edx-platform/pull/25695/commits/6fa07ec6c7863885a1f9f80ab3370001e37698f3)

The reason I took this approach, instead of "stringifying" the JSON object received by webpack, is that I ran into problems with Single and Double quotes again 🙁 
That means that I'll need to start replacing the single quotes with double quotes, as was previously done: `replace("'", '"')`.

What are your thoughts on this?

bradenmacdonald replied:
> Unfortunately, that's due to the print_setting that is being used to extract the settings from CMS and LMS.

This seems like the root of the problem - it's printing in python format, not JSON. Perhaps there is something else we can use that will output the setting as JSON in the first place, instead of having to reformat it from python to JSON?

bradenmacdonald replied:
i.e. can we modify https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/util/management/commands/print_setting.py to give it a `--json` option, to format the setting value as JSON instead of just printing it as python?

**or**, what's probably much easier, is you can use the `derived` helper to create a derived setting that's string-encoded as JSON. See this example of a derived setting:

https://github.com/edx/edx-platform/blob/3d79b8ab95c0e97c4657528fd303db57196d0c4f/cms/envs/common.py#L590-L591

So then you can go like this:

```python
JS_ENV_EXTRA_CONFIG = {}
# Derived setting which is just the above, JSON-encoded into a string; do not change this directly:
JS_ENV_EXTRA_CONFIG_JSON = lambda settings: json.dumps(settings.JS_ENV_EXTRA_CONFIG)
derived('JS_ENV_EXTRA_CONFIG_JSON')
```

nizarmah replied:
I love how easy the second solution is, and I was considering it. But then I realized that there might be some confusion with variable names.

So, I ended up adding a `--json` option to the `print_setting` command. I'd like to think I did a decent job with it.

I do suspect some comments on the changes though, especially that I might have added an unnecessary/unneeded `get_django_json_settings`, which might make its docstring outdated in the future...

bradenmacdonald replied:
Nice, thanks - that looks great, and I think it is a better approach :)

nizarmah commented:
@bradenmacdonald this is ready for your review again. 

Sorry about the problem you faced earlier! I added some explanation about it in the changes you requested 👍🏼 

Please note that there seems to be a problem with one of the builds on the whole edX platform, so the currently failing python test isn't a result of the changes I did in this pull request.

bradenmacdonald commented on pavelib/utils/envs.py:
```diff
@@ -271,6 +274,22 @@ def get_django_settings(cls, django_settings, system, settings=None):
+        :param django_settings: list of django settings values to get
+        :param system: the django app to use when asking for the setting (lms | cms)
+        :param settings: the settings file to use when asking for the value
+        :return: json value of the django setting
```
Does this return a string or a decoded JSON object?

nizarmah replied:
This returns a string, since it changes the `True` values to `true` which would be invalid in a Python object.

I'll update the docstring to mention it in a better way. 👍🏼

nizarmah replied:
Addressed 👍🏼 

Updated to 
```
        :return: json string value of the django setting
```

bradenmacdonald commented on pavelib/assets.py:
```diff
@@ -776,24 +775,12 @@ def webpack(options):
-    js_env_extra_config_setting, = Env.get_django_settings(["JS_ENV_EXTRA_CONFIG"], "cms", settings=settings)
-    js_env_extra_config = json.dumps(json_format_setting(js_env_extra_config_setting) or {})
+    js_env_extra_config_setting, = Env.get_django_json_settings(["JS_ENV_EXTRA_CONFIG"], "cms", settings=settings)
+    js_env_extra_config = json.dumps(js_env_extra_config_setting or {})
```
I think that here `js_env_extra_config_setting` is a string right? So `"string" or {}` seems like inconsistent typing - should it be `js_env_extra_config_setting or "{}"` ? Or is `js_env_extra_config_setting` an object already?

nizarmah replied:
`js_env_extra_config_setting` is a string, yes.

You're right, it is inconsistent. I'll update this to `js_env_extra_config_setting or "{}"`. In addition, I'll update the tests too, I added comments to address those changes 👍🏼

nizarmah replied:
Addressed 👍🏼

bradenmacdonald reviewed (approved):
Thanks for your patience and diligence here, refining this approach and documenting it!

:+1: if you address these last questions, rebase and squash into a single commit with a clear commit message.

Please also do a quick test to make sure that this doesn't introduce any problems when the setting is not set (i.e. default values everywhere).

- [x] I tested this: on the sandbox
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation

nizarmah commented on pavelib/utils/envs.py:
```diff
@@ -271,6 +274,22 @@ def get_django_settings(cls, django_settings, system, settings=None):
+            django_settings,
+            system,
+            settings=settings,
+            print_setting_args=["--json"]
```
Add trailing comma here please

nizarmah replied:
Addressed 👍🏼

nizarmah commented on pavelib/paver_tests/test_servers.py:
```diff
@@ -297,6 +302,7 @@ def verify_run_all_servers_task(self, options):
                 node_env="production",
                 static_root_lms=None,
                 static_root_cms=None,
+                js_env_extra_config=json.dumps({}),
```
According to the [comment raised earlier](https://github.com/edx/edx-platform/pull/25695/files#r553657732), update the `json.dumps({})` to `json.dumps("{}")` for consistency 👍🏼

nizarmah replied:
Addressed 👍🏼

nizarmah commented on pavelib/paver_tests/test_servers.py:
```diff
@@ -251,6 +255,7 @@ def verify_server_task(self, task_name, options, contracts_default=False):
                 node_env="production",
                 static_root_lms=None,
                 static_root_cms=None,
+                js_env_extra_config=json.dumps({}),
```
According to the [comment raised earlier](https://github.com/edx/edx-platform/pull/25695/files#r553657732), update the `json.dumps({})` to `json.dumps("{}")` for consistency 👍🏼

nizarmah replied:
Addressed 👍🏼

nizarmah commented:
@bradenmacdonald It's my pleasure! Your comments are something I look forward to, because I always end up learning something new, whether about the edX platform or on a personal level 😃 

I addressed the comments you raised, and I also squashed the commits.

Regarding testings, I deployed two new sandboxes:
- No TinyMCE plugins (default settings/configuration)
  - LMS: https://configpr6179.sandbox.opencraft.hosting/
  - Studio: https://studio.configpr6179.sandbox.opencraft.hosting/
- Custom TinyMCE plugin
  - LMS: https://pr25695.sandbox.opencraft.hosting/
  - Studio: https://studio.configpr6179.sandbox.opencraft.hosting/

I have tested both sandboxes, and the TinyMCE editors are running great on both 👍🏼 
Please note that the sandboxes are running the following:
- [open-craft/edx-platform:nizar/tinymce_plugins_combined_dependencies_master](https://github.com/open-craft/edx-platform/tree/nizar/tinymce_plugins_combined_dependencies_master)
  - Identical to this branch, but cherry-picks the change from https://github.com/edx/edx-platform/pull/25324
- [edx/configuration:master](https://github.com/edx/configuration/)
  - Includes the changes from https://github.com/edx/configuration/pull/6179

edx-webhook commented:
@nizarmah 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/25695&entry.1484655318=https://github.com/nizarmah&entry.752974735=2020-11-29+16:35&entry.1917517419=2021-01-11+16:28&entry.2133058324=Yes) so we can improve your experience in the future.
